Fetch order by bug

jverswijver opened the pull request:
Added test based off of linked issue. It seems enforcing DISTINCT when we are making sql queries with primary keys was not useful, I have removed it for now.

fix #914

guzman-raphael commented on tests/test_fetch.py:
```diff
@@ -287,3 +288,8 @@ def test_query_caching(self):
+    def test_fetch_group_by(self):
+        # https://github.com/datajoint/datajoint-python/issues/914
+
+        Parent().fetch('KEY', order_by='name')
```
Would you add an `assert` statement here to be specific on what we are expecting from this query used in this example?

jverswijver replied:
addressed this in commit ce16d0e

guzman-raphael commented on datajoint/expression.py:
```diff
@@ -106,9 +106,7 @@ def make_sql(self, fields=None):
-        distinct = self.heading.names == self.primary_key
-        return 'SELECT {distinct}{fields} FROM {from_}{where}'.format(
-            distinct="DISTINCT " if distinct else "",
+        return 'SELECT {fields} FROM {from_}{where}'.format(
```
Can you also explore all usage of `DISTINCT`. Noticed that there is another similar usage in the `Aggregation` class. Did you mean to remove its use here but maintain its use there?

jverswijver replied:
I looked through it a bit and saw that it also uses `DISTINCT`, but it might be best to just leave it until a user reports a specific issue that is related to the aggregation usage of distinct. It would be a bad idea to remove it without a specific reason/issue.

guzman-raphael reviewed (changes requested):
Thanks for the PR @jverswijver! Primarily just have some questions about other similar usage of `DISTINCT` in source.

dimitri-yatsenko commented on tests/test_fetch.py:
```diff
@@ -287,3 +288,12 @@ def test_query_caching(self):
+
+        try:
+            Parent().fetch('KEY', order_by='name')
+            assert True
```
This has no effect.

dimitri-yatsenko commented on tests/test_fetch.py:
```diff
@@ -287,3 +288,12 @@ def test_query_caching(self):
+    def test_fetch_group_by(self):
+        # https://github.com/datajoint/datajoint-python/issues/914
+
+        try:
```
This `try` ... `except` does not appear to be necessary. Just letting the exception raise will have the same effect.

dimitri-yatsenko commented:
`DISTINCT` becomes important in queries of the following types:
```python
dj.U('attr') & q
```
where `attr` is not the primary key of `q`.

Another case is unions. 
```python
q + r
```

If we don't have test cases that require `DISTINCT`, let's add them

jverswijver commented:
A few updates:

- github actions are failing because an api we use for code coverage is down for maintenance but the code up to commit 13b77d4 passes the test suite on my machine.
- I have added an additional test that ensures that the results of dj.U operations are unique.
- I have added the _distinct property to the QueryExpression class so we can be explicit when we want to use `DISTINCT` in our query formulation.
- I have updated the U class to use this new _distinct property to ensure that the results produced are unique like they were before when we blanket applied `DISTINCT`.

guzman-raphael commented on tests/test_fetch.py:
```diff
@@ -287,3 +288,41 @@ def test_query_caching(self):
+        print(fetched_data)
+        assert fetched_data == expected_data
+
+    def test_dj_U_DISTINCT(self):
```
Following python's `snake_case` convention for functional naming.

```suggestion
    def test_dj_u_distinct(self):
```

Otherwise, a great test. :+1:

guzman-raphael commented on datajoint/expression.py:
```diff
@@ -101,14 +104,13 @@ def where_clause(self):
             str(s) for s in self.restriction)
 
-    def make_sql(self, fields=None):
+    def make_sql(self, fields=None, distinct=True):
```
Hmm, don't think this is serving any purpose now, right?

```suggestion
    def make_sql(self, fields=None):
```

guzman-raphael reviewed (changes requested):
@jverswijver Nice work in determining the root cause for this one! :muscle:
Only providing some minor feedback here.

guzman-raphael reviewed (approved):
@jverswijver Nice job man! :chart_with_upwards_trend:
